Add subsetting of the datasets

TheMrSheldon opened the pull request:
Sometimes one may just want to train, evaluate or test on a smaller subset of the dataset (e.g., for testing or to see if the model is able to overfit on this smaller portion).

This feature is implement by adding three new parameters to the `H5DataModule`: `limit_train_set`, `limit_val_set`, and `limit_test_set`. They allow for no subsetting, using a fixed number of entries, or using a portion of the full dataset.
All three parameters are optional and the default behavior is unchanged (i.e, no subsetting).

mrjleo commented:
Hi @TheMrSheldon, thanks a lot for the PR.
I have one question though: There is a similar functionality in PyTorch Lightning (e.g. [limit_train_batches](https://pytorch-lightning.readthedocs.io/en/1.8.3/common/trainer.html)). Is there any difference other than the fact that your implementation limits instances and not batches?

TheMrSheldon commented:
Before implementing `limit_train_set` I had actually used `limit_train_batches`. The key difference is however that the data of these limited number of batches are still drawn from the entire dataset. To give an illustrative example:
 The dataset {A, B, C, D, E, F, G} may be limited to 2 batches (of size 2) such that the model trains on {{A, C}, {F, G}} in epoch 1 and on {{B, D}, {A, E}} on epoch 2 and thus would still train on the entire dataset while only seeing 2 batches per epoch.

With `limit_train_set` in particular, I would like to limit the dataset such that I could try to overfit with the model only on a small subset of the dataset which is not possible with `limit_train_batches`.

mrjleo commented on ranking_utils/model/data/h5.py:
```diff
@@ -251,6 +254,14 @@ def __init__(
+            limit_train_set (Union[int, float], optional): loads only a subset of the training set. If an integer is
+                passed, it indicates that only the first `limit_train_set` entries in the training set should be used.
+                If a float is parsed, it indicates that only the first `|train_set|*limit_train_set` entries in the
+                training set should be used. E.g., train_set=100 indicates to train on 100 dataset entries (or less if
```
I think this should be `limit_train_set=100`

mrjleo commented:
I see, that is indeed a nice addition. It would be great if you could do the following:
- Format your code using [isort](https://pycqa.github.io/isort/) (first) and [black](https://github.com/psf/black) (second)
- Fix the docstring (see comment)

Thanks again!

TheMrSheldon commented:
Will do. Thank you very much for the feedback. I was also thinking if `_subset_dataset` should better be positioned outside the class since it does not need to be a member and strictly speaking is semantically unrelated to `H5DataModule`. Which would you prefer?

mrjleo commented:
Good point. Maybe it could go directly to `ranking_utils.model.data`?

TheMrSheldon commented:
All done. I moved `_subset_dataset` to `ranking_utils.model.data` as you suggested. I also removed the initial underscore since I would now consider it a public utility function.

I feel like the model and data packets should probably be split into more subfiles and `ranking_utils.model.data` is quite confusing since it should maybe be its own submodule of `ranking_utils`: `ranking_utils.data`.

Since that would be out of scope for this ticket I would however do that with another pull request if you agree with my assessment and would consider this ticket closed if you are fine with the last changes.

mrjleo commented on ranking_utils/model/data/h5.py:
```diff
@@ -251,6 +256,14 @@ def __init__(
+                passed, it indicates that only the first `limit_train_set` entries in the training set should be used.
+                If a float is parsed, it indicates that only the first `|train_set|*limit_train_set` entries in the
+                training set should be used. E.g., limit_train_set=100 indicates to train on 100 dataset entries (or
+                less if the dataset does not have enough entries), whereas train_set=.5 indicates to train on 50% of the
```
Should be `limit_train_set=.5`

mrjleo commented:
> All done. I moved `_subset_dataset` to `ranking_utils.model.data` as you suggested. I also removed the initial underscore since I would now consider it a public utility function.

Great, thanks! There's one more type, afterwards it can be merged.

> I feel like the model and data packets should probably be split into more subfiles and `ranking_utils.model.data` is quite confusing since it should maybe be its own submodule of `ranking_utils`: `ranking_utils.data`

That's true. However, `ranking_utils.datasets` already exists, so this might be confusing as well. Also, it'll break the API. I'll have to think about it.

TheMrSheldon commented:
> That's true. However, `ranking_utils.datasets` already exists, so this might be confusing as well.

I believe, it could be reasoned well that `ranking_utils.data` is (similarly to `torch.utils.data`) simply containing definitions on how data is structured and functions relating to this structure (one of those structures being a datasets and possible subclasses) whereas `ranking_utils.datasets` (similarly to `torchvision.datasets`) contains only concrete Dataset-definitions.

> Also, it'll break the API. I'll have to think about it.

At first one could use the `__init__.py` such that both `ranking_utils.data` and `ranking_utils.model.data` would work to keep some backwards compatibility but also print warnings when `ranking_utils.model.data` is loaded. Tags in git would further allow for versioning and at some point, with a major version change, `ranking_utils.mode.data` could be removed
